src/core_plugins/vega/public/vega_editor_controller.js

```diff
@@ -0,0 +1,66 @@
+import compactStringify from 'json-stringify-pretty-compact';
+import hjson from 'hjson';
+import { Notifier } from 'ui/notify';
+import { uiModules } from 'ui/modules';
+
+const module = uiModules.get('kibana/vega', ['kibana']);
+module.controller('VegaEditorController', ($scope /*, kbnUiAceKeyboardModeService*/) => {
+
+  const notify = new Notifier({ location: 'Vega' });
+
+  return new (class VegaEditorController {
+    constructor() {
+      $scope.aceLoaded = (editor) => {
+        editor.$blockScrolling = Infinity;
+
+        const session = editor.getSession();
+        session.setTabSize(2);
+        session.setUseSoftTabs(true);
+
+        this.aceEditor = editor;
+      };
+
+      $scope.formatJson = (event) => {
+        this._format(event, compactStringify, {
+          maxLength: this._getCodeWidth(),
+        });
+      };
+
+      $scope.formatHJson = (event) => {
+        this._format(event, hjson.stringify, {
+          condense: this._getCodeWidth(),
+          bracesSameLine: true,
+          keepWsc: true,
+        });
+      };
+    }
+
+    _getCodeWidth() {
+      return this.aceEditor.getSession().getWrapLimit();
+    }
+
+    _format(event, stringify, opts) {
+      event.preventDefault();
+
+      let newSpec;
+      try {
+        const spec = hjson.parse(this.aceEditor.getSession().doc.getValue(), { legacyRoot: false, keepWsc: true });
```

Comment:
Why not use the `ng-model` variable here: `$scope.vis.params.spec`?

Reply:
Per below, keeping access to the editor state consistent - both getter and setter are used on the same ACE editor object.